### PR TITLE
BENCH/REF: parametrize CSV benchmarks on engine

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

The diff in basic PR implementing the `pyarrow`-based CSV engine (#38370) is quite big. A part of that PR is a small refactor of the CSV I/O benchmarks such that they take `engine` as a parameter. Most of that refactor is not dependent on having the pyarrow engine so I'm spinning it off here to de-bloat #38370.

### CODE REVIEW

**Comment by jreback** [comment]
looks fine, checks are failing though, ping on green.

cc @gfyoung @WillAyd if comments.

**Comment by arw2019** [comment]
Atm not sure what's going on with the benchmark failure  - the failure is in `asv_bench/benchmarks/tslibs/timedelta.py` which doesn't get touched here

**Comment by mroeschke** [comment]
https://github.com/pandas-dev/pandas/pull/38476 should fix the unrelated benchmark failure.

File: asv_bench/benchmarks/io/csv.py [hunk 1]
```diff
@@ -146,10 +146,10 @@ def time_read_csv(self, bad_date_value):
 class ReadCSVSkipRows(BaseIO):
 
     fname = "__test__.csv"
-    params = [None, 10000]
-    param_names = ["skiprows"]
+    params = ([None, 10000], ["c"])
```
Comment (WillAyd) [inline asv_bench/benchmarks/io/csv.py:149]: Is this supposed to be c and python or just c?

**Review by WillAyd — COMMENTED** [review]
looks good small question

File: asv_bench/benchmarks/io/csv.py (hunk 1 above)
Comment (arw2019) [reply asv_bench/benchmarks/io/csv.py:149]: On current master we only have the default (I think that's `c`)

We have `c` and `python` in all the other benchmarks so it seems reasonable to add that here. If there's a reason we don't want that I'll revert

**Comment by arw2019** [comment]
> #38476 should fix the unrelated benchmark failure.

Thanks @mroeschke!!

File: asv_bench/benchmarks/io/csv.py [hunk 2]
```diff
@@ -255,8 +255,24 @@ def time_read_csv_python_engine(self, sep, decimal, float_precision):
...
+
+    def time_read_stringcsv(self, engine):
+        read_csv(self.data(self.StringIO_input), engine=engine)
+
+    def time_read_bytescsv(self, engine):
+        read_csv(self.data(self.BytesIO_input), engine=engine)
 
+
+class ReadCSVCategorical(BaseIO):
     fname = "__test__.csv"
```
Comment (jorisvandenbossche) [inline asv_bench/benchmarks/io/csv.py:277]: could also parametrize this one?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks!

Is there some logic in which classes are parameterized and which not?

File: asv_bench/benchmarks/io/csv.py (hunk 2 above)
Comment (arw2019) [reply asv_bench/benchmarks/io/csv.py:280]: Done.

**Comment by arw2019** [comment]
> Is there some logic in which classes are parameterized and which not?

Not that I know of beyond (AFAIK) that the python engine supports all these whereas the c/pyarrow engines only support a proper subset. 

I'll do a pass through the entire file and parametrize where both c and python engines are supported.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by arw2019** [comment]
CI green

**Comment by jreback** [comment]
thanks @arw2019

**Pull request merged**